#KNN-->K-Nearest Neibhours

--->It is a supervised Learning Algorithm which is uesd for  Regression and classification both.
--->Regression---->We sugest this,when target column is continuous values.This predict output by taking avg.
--->classification--->we sugest this, when target column is decide through majarity.for example YES/NO,SPAM/NOT SPAM,PASS/FAIL.

-->This algarithm uses distance formula(Ecluidian/manhatten)

          Distance formula:Ecluidian   : sqrt((x2-x1)^2 +(y2-y1)^2)
          
                           Manhattan Distance:|x2-x1| +|y2-y1| ---->this is uses when data is more complex we use thhis formula

-->It is lazy algorithm,during training it stores the data.

---->We need to use scalling techniques in this because we are using distance formula.

-->k-values are given by using sqrt(n). here n--->refers to total samples

-->try to avoid giving n value even to avoid tie braker.for example :Classification algorithm dicides by using majarity when n even  if more chances to get tie and the model fails to predict correct output sometimes.

Example:PASS=2 ,FAIL=2---->HERE MODEL PICKS ONE OF IT.

#parameters
1)n_neighbors=sqrt(n)
2)metric="elcludies" | "mamhattan"

Advantages:

1)Easy to understand
2)Easy to implement
3)No training phase
4)Works well on small datasets
5)Can solve both classification and regression problems


Disadvantages:

1)Slow for large datasets
2)Requires feature scaling
3)Sensitive to noisy data
4)Choosing the right K is important
5)Performance decreases with many features (curse of dimensionality)


In [6]:
from sklearn.neighbors import KNeighborsClassifier, KNeighborsRegressor
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import LabelEncoder,StandardScaler
from sklearn.metrics import accuracy_score,confusion_matrix,classification_report
import pandas as pd



In [7]:
df=pd.read_csv('retail_customer_segmentation.csv')
df.head()

,customer_id,age,annual_income,months_active,avg_monthly_spend,purchase_frequency,avg_order_value,discount_usage_rate,return_rate,browsing_time_minutes,support_interactions,payment_method,region,customer_segment
0,33554,53,100473.211709,63,121.430565,0.817268,66.820403,0.117256,0.023144,77.298393,2.0,Card,Semi-Urban,Occasional
1,9428,54,54730.644845,67,572.552674,3.176551,137.087449,0.261647,0.429054,92.132565,2.0,Wallet,Urban,Occasional
2,200,44,58268.121079,57,266.593896,2.713168,71.796888,0.284785,0.011854,155.194768,1.0,UPI,Rural,Occasional
3,12448,54,64829.795654,40,691.452358,5.553977,105.501185,0.104832,0.399686,113.917756,0.0,Wallet,Rural,High_Value
4,39490,28,27431.467873,15,832.664792,1.348389,354.568534,0.409204,0.039517,50.123656,1.0,Card,Semi-Urban,Occasional


In [8]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 14 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   customer_id            50000 non-null  int64  
 1   age                    50000 non-null  int64  
 2   annual_income          46925 non-null  float64
 3   months_active          50000 non-null  int64  
 4   avg_monthly_spend      47480 non-null  float64
 5   purchase_frequency     48021 non-null  float64
 6   avg_order_value        50000 non-null  float64
 7   discount_usage_rate    47451 non-null  float64
 8   return_rate            47513 non-null  float64
 9   browsing_time_minutes  46066 non-null  float64
 10  support_interactions   48012 non-null  float64
 11  payment_method         50000 non-null  object 
 12  region                 50000 non-null  object 
 13  customer_segment       50000 non-null  object 
dtypes: float64(8), int64(3), object(3)
memory usage: 5.3+ 

In [9]:
#check null values
df.isnull().sum()

customer_id                 0
age                         0
annual_income            3075
months_active               0
avg_monthly_spend        2520
purchase_frequency       1979
avg_order_value             0
discount_usage_rate      2549
return_rate              2487
browsing_time_minutes    3934
support_interactions     1988
payment_method              0
region                      0
customer_segment            0
dtype: int64

In [10]:
#this for single column
df["annual_income"]=df["annual_income"].fillna(df["annual_income"].median())

In [11]:
df.isnull().sum()

customer_id                 0
age                         0
annual_income               0
months_active               0
avg_monthly_spend        2520
purchase_frequency       1979
avg_order_value             0
discount_usage_rate      2549
return_rate              2487
browsing_time_minutes    3934
support_interactions     1988
payment_method              0
region                      0
customer_segment            0
dtype: int64

In [12]:
#this for multiple columns  it replaces througth median values with NaN
df.fillna(df.median(numeric_only=True),inplace=True)

In [13]:
df.isnull().sum()

customer_id              0
age                      0
annual_income            0
months_active            0
avg_monthly_spend        0
purchase_frequency       0
avg_order_value          0
discount_usage_rate      0
return_rate              0
browsing_time_minutes    0
support_interactions     0
payment_method           0
region                   0
customer_segment         0
dtype: int64

In [14]:
#separate the data into features and target

x=df.drop("customer_segment",axis=1)

#use label encoding for target col
le=LabelEncoder()
y=le.fit_transform(df["customer_segment"])


In [15]:
x.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 13 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   customer_id            50000 non-null  int64  
 1   age                    50000 non-null  int64  
 2   annual_income          50000 non-null  float64
 3   months_active          50000 non-null  int64  
 4   avg_monthly_spend      50000 non-null  float64
 5   purchase_frequency     50000 non-null  float64
 6   avg_order_value        50000 non-null  float64
 7   discount_usage_rate    50000 non-null  float64
 8   return_rate            50000 non-null  float64
 9   browsing_time_minutes  50000 non-null  float64
 10  support_interactions   50000 non-null  float64
 11  payment_method         50000 non-null  object 
 12  region                 50000 non-null  object 
dtypes: float64(8), int64(3), object(2)
memory usage: 5.0+ MB


Here Still after clearing null values and separeated feature and target  and features still we have object columns.we need to convert to numuric values 

-here unique values are 3 in both we can use LabelEncoder

In [17]:
#we create separete obj for labelencoser
f_le=LabelEncoder()
obj_col=x.select_dtypes(include=object).columns
for col in obj_col:
    x[col]=f_le.fit_transform(x[col])

In [19]:
x.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 13 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   customer_id            50000 non-null  int64  
 1   age                    50000 non-null  int64  
 2   annual_income          50000 non-null  float64
 3   months_active          50000 non-null  int64  
 4   avg_monthly_spend      50000 non-null  float64
 5   purchase_frequency     50000 non-null  float64
 6   avg_order_value        50000 non-null  float64
 7   discount_usage_rate    50000 non-null  float64
 8   return_rate            50000 non-null  float64
 9   browsing_time_minutes  50000 non-null  float64
 10  support_interactions   50000 non-null  float64
 11  payment_method         50000 non-null  int32  
 12  region                 50000 non-null  int32  
dtypes: float64(8), int32(2), int64(3)
memory usage: 4.6 MB


#Now every column is integer

In [28]:
#split the data 
xtrain,xtest,ytrain,ytest=train_test_split(x,y,test_size=0.2,random_state=42)
xtrain

,customer_id,age,annual_income,months_active,avg_monthly_spend,purchase_frequency,avg_order_value,discount_usage_rate,return_rate,browsing_time_minutes,support_interactions,payment_method,region
39087,7359,64,31049.037698,50,303.810997,7.778965,34.606699,0.499974,0.044472,39.056486,1.0,1,0
30893,48517,27,34902.671586,39,297.477404,13.896417,12.190045,0.142636,0.404013,15.591307,0.0,0,0
45278,19745,21,21739.637289,49,284.156218,3.900793,57.981683,0.198487,0.018420,87.606898,1.0,0,0
16398,46726,32,112497.680559,58,513.286273,12.367122,38.399162,0.217201,0.148782,25.143615,1.0,0,0
13653,30412,49,66417.416926,19,423.441877,3.109467,103.040580,0.132811,0.288048,79.736351,1.0,2,0
...,...,...,...,...,...,...,...,...,...,...,...,...,...
11284,3319,38,7889.513019,53,357.829203,4.098220,70.187089,0.482523,0.038928,17.859021,0.0,2,1
44732,26376,63,11975.072693,70,297.477404,2.413184,71.898669,0.190498,0.413438,22.503877,3.0,2,2
38158,44095,62,19640.312878,21,82.334664,3.856241,16.954402,0.207793,0.086062,40.866841,3.0,0,2
860,26038,38,25069.280484,33,225.388165,4.419216,41.590546,0.231475,0.059456,98.884253,3.0,0,0


In [31]:
#use scalling tecqniques-->StandardScaler 

scaler=StandardScaler()
#for train data
x_train_updated=scaler.fit_transform(xtrain)
#for test data we use only transform  
x_test_updated=scaler.transform(xtest)


In [34]:
# #creates model for KNN
# for k in range(1,5):
#     model=KNeighborsClassifier(n_neighbors=k)
#     model.fit(x_train_updated,ytrain)

#     y_pred_train=model.predict(x_train_updated)

#     print(f" For k :{k} Accuracy Score:{accuracy_score(ytrain,y_pred_train)}")
    


In [38]:
model=KNeighborsClassifier(n_neighbors=5)
model.fit(x_train_updated,ytrain)

y_pred_train=model.predict(x_train_updated)

print(f" For k :{5} Accuracy Score:{accuracy_score(ytrain,y_pred_train)}")


 For k :5 Accuracy Score:0.748475


In [44]:
print(confusion_matrix(ytrain,y_pred_train))

[[ 3052   700   332   140]
 [  460  5614   559   645]
 [  167   622 15232  1741]
 [  180  1098  3417  6041]]


In [52]:
print(classification_report(ytrain,y_pred_train, target_names=le.classes_))

              precision    recall  f1-score   support

  High_Value       0.79      0.72      0.76      4224
       Loyal       0.70      0.77      0.73      7278
  Occasional       0.78      0.86      0.82     17762
     Regular       0.71      0.56      0.63     10736

    accuracy                           0.75     40000
   macro avg       0.74      0.73      0.73     40000
weighted avg       0.75      0.75      0.74     40000



In [60]:
#Model evaluation for test data

y_test_pred=model.predict(x_test_updated)

print("Accuracy Score for train data:",accuracy_score(ytest,y_test_pred))

Accuracy Score for train data: 0.6212


In [62]:
print(confusion_matrix(ytest,y_test_pred))

[[ 593  269  129   58]
 [ 173 1220  210  282]
 [  58  222 3309  758]
 [  59  412 1158 1090]]


In [64]:
print(classification_report(ytest,y_test_pred, target_names=le.classes_))

              precision    recall  f1-score   support

  High_Value       0.67      0.57      0.61      1049
       Loyal       0.57      0.65      0.61      1885
  Occasional       0.69      0.76      0.72      4347
     Regular       0.50      0.40      0.44      2719

    accuracy                           0.62     10000
   macro avg       0.61      0.59      0.60     10000
weighted avg       0.61      0.62      0.61     10000



Conclusion

The KNN classifier achieved an accuracy of 62.12% on the test dataset. Among all customer segments, Occasional customers were classified most accurately, achieving an F1-score of 0.72 and a recall of 0.76, indicating that the model successfully identified most customers in this segment. The performance for High_Value and Loyal customers was moderate, with F1-scores of 0.61 each. However, the model struggled to classify Regular customers, resulting in a lower precision of 0.50, recall of 0.40, and F1-score of 0.44. Overall, the model provides moderate classification performance, but improving feature engineering or tuning the value of K may enhance the prediction accuracy, especially for the Regular customer segment.